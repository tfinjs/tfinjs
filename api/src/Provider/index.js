import JsToHcl from '../JsToHcl';

/**
 * Creates a new provider instance for an Api
 *
 * @param {string} provider name
 * @param {object} body of the provider config
 * @param {providerUri} providerUri
 * @class Provider
 */
class Provider {
  constructor(provider, body, uri) {
    this.provider = provider;
    this.body = body;
    this.uri = uri;
  }

  /**
   * Get the providerUri
   *
   * @returns {providerUri} providerUri
   * @memberof Provider
   */
  getUri() {
    return this.uri;
  }

  /**
   * Gets the HCL for the provider
   *
   * @returns {hcl} The provider code
   * @memberof Provider
   */
  getHcl() {
    const converter = new JsToHcl();
    const providerBody = converter.stringify(this.body);
    return `provider "${this.provider}" { ${providerBody} }`;
  }
}

export default Provider;
