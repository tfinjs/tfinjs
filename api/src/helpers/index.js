import assert from 'assert';
import Resource from '../Resource';
import createTerraformStringInterpolation from '../statics/createTerraformStringInterpolation';
import { TERRAFORM_OUTPUT_PREFIX } from '../constants';

/**
 * Factory function to a function where the returned function takes a resource and returns the versionedName of the resource.
 *
 * @returns {function} versionedName
 */
export const versionedName = () => (resource) => {
  assert(
    resource instanceof Resource,
    'resource must be an instance of Resource',
  );
  return resource.versionedName();
};

/**
 * Factory function to a function which takes a resource and returns the terraform interpolation value which references the key
 *
 * @param {resource} resource - a resource
 * @param {string} key - the key referenced on the resource
 * @returns {function} reference - gets the reference string of the resource
 */
export const reference = (resource, key) => {
  assert(
    resource instanceof Resource,
    'resource must be an instance of Resource',
  );
  assert(typeof key === 'string', 'key must be string');

  return (referencedResource) => {
    assert(
      referencedResource instanceof Resource,
      'referencedResource must be an instance of Resource',
    );
    assert(
      referencedResource !== resource,
      'you cannot reference the resource itself',
    );
    referencedResource.registerRemoteState(resource);
    resource.addOutputKey(key);

    /* prefix because of https://github.com/hashicorp/terraform/issues/7982 */
    return createTerraformStringInterpolation(
      `data.terraform_remote_state.${resource.versionedName()}.${TERRAFORM_OUTPUT_PREFIX}${key}`,
    );
  };
};
