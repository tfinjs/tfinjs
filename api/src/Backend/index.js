import assert from 'assert';
import throwError from '../statics/throwError';
import JsToHcl from '../JsToHcl';
import requiredParam from '../statics/requiredParam';
import Provider from '../Provider';
import DeploymentConfig from '../DeploymentConfig';
import Resource from '../Resource';

/**
 * Creates an instance of Backend.
 *
 * @param {string} backend - the backend name
 * @param {object} config
 * @param {function=} config.backendConfig - Config for the terraform { backend { ... } }
 * @param {function=} config.dataConfig - Config for the terraform_remote_state
 * @param {provider=} config.provider - The provider which will be used to deploy the backend source
 * @param {function=} config.create - A function taking resource creating function as the first argument and expects the resource to be returned. The resource should be the backend source.
 * @class Backend
 */
class Backend {
  constructor(
    backend = requiredParam('backend'),
    {
      backendConfig, dataConfig, provider, create,
    } = {},
  ) {
    assert(
      typeof backend === 'string' || backend === null,
      'backend must be string or null',
    );
    assert(
      typeof backendConfig === 'undefined'
        || typeof backendConfig === 'function',
      'backendConfig must be function if defined',
    );
    assert(
      typeof dataConfig === 'undefined' || typeof dataConfig === 'function',
      'dataConfig must be function if defined',
    );
    assert(
      typeof create === 'function' || typeof create === 'undefined',
      'create must be function if defined',
    );
    assert(
      typeof provider === 'undefined' || provider instanceof Provider,
      'provider must be instance of Provider if defined',
    );

    if ((provider && !create) || (!provider && create)) {
      throwError(
        'You must provide both the provider and the create function',
        this.constructor,
      );
    }

    this.backend = backend;
    this.backendConfig = backendConfig;
    this.dataConfig = dataConfig;

    this.provider = provider;
    this.privateCreate = create;
  }

  /**
   * If the backend should be created, (only if provider and create has been provided)
   *
   * @returns {boolean} shouldCreateBackend - If the backend should be created
   * @memberof Backend
   */
  shouldCreateBackend() {
    return this.provider && this.create;
  }

  /**
   * Gets the provided provider
   *
   * @returns {provider} provider - Instance of provider optionally provided to the constructor
   * @memberof Backend
   */
  getProvider() {
    return this.provider;
  }

  /**
   * Gets the create function which will create the resource that constitutes the backend
   *
   * @returns {function} create - The create function optionally provided to the contructor
   * @memberof Backend
   */
  create(deploymentConfig) {
    assert(
      deploymentConfig instanceof DeploymentConfig,
      'deploymentConfig must be an instance of DeploymentConfig, in Backend.create',
    );
    const result = this.privateCreate(deploymentConfig);
    assert(
      result instanceof Resource,
      'You must return an instance of Resource',
    );
    return result;
  }

  /**
   * Creates the terraform backend configuration in HCL
   *
   * @param {versionedName} versionedName
   * @returns {hcl} hcl
   * @memberof Backend
   */
  getBackendHcl(versionedName = requiredParam('versionedName')) {
    assert(typeof versionedName === 'string', 'versionedName must be string');

    if (this.backend === null) {
      return '';
    }

    assert(
      this.backendConfig,
      'You must define the backendConfig on the backend in order to make other resources be able to reference the resource (unless the specified backend is null)',
    );

    const jsToHcl = new JsToHcl();
    const hcl = `terraform { backend "${this.backend}" { ${jsToHcl.stringify(
      this.backendConfig(versionedName),
    )} } }`;

    return hcl;
  }

  /**
   * Creates the terraform_remote_state configuration in HCL
   *
   * @param {versionedName} versionedName
   * @returns {hcl} hcl
   * @memberof Backend
   */
  getDataConfig(versionedName = requiredParam('versionedName')) {
    assert(typeof versionedName === 'string', 'versionedName must be string');
    assert(
      this.dataConfig,
      'You must define the dataConfig on the backend in order to be able to reference other resources',
    );

    const jsToHcl = new JsToHcl();
    const params = {
      backend: 's3',
      config: this.dataConfig(versionedName),
    };
    const hcl = `data "terraform_remote_state" "${versionedName}" {${jsToHcl.stringify(
      params,
    )}}`;

    return hcl;
  }
}
export default Backend;
