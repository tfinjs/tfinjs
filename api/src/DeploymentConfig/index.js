import assert from 'assert';
import Namespace from '../Namespace/NamespaceParent';
import Provider from '../Provider';
import DeploymentConfigParent from './DeploymentConfigParent';

/**
 * Creates an instance of DeploymentConfig.
 *
 * @param {namespace} namespace - An instance of the Namespace class
 * @param {object} config - Configuration object for the deployment config
 * @param {string} config.environment - The deployment environment
 * @param {string} config.version - The deployment version
 * @param {provider} config.provider - And instance of the Provider class which defines to which tennant and location resources with this config should be deployed
 * @class DeploymentConfig
 */
class DeploymentConfig extends DeploymentConfigParent {
  constructor(namespace, { environment, version, provider }) {
    super();
    assert(
      namespace instanceof Namespace,
      'namespace must be an instance of Namespace',
    );
    assert(typeof environment === 'string', 'environment must be string');
    assert(typeof version === 'string', 'version must be string');
    assert(
      provider instanceof Provider,
      'provider must be an instance of Provider',
    );

    this.namespace = namespace;
    this.environment = environment;
    this.version = version;
    this.provider = provider;
  }

  /**
   * Gets the uri of the API.
   * Is unique based on
   * project, environment, version,
   * providerUri and the namespace.
   *
   * @returns {apiUri} apiUri - The Api uri
   * @memberof DeploymentConfig
   */
  getUri() {
    const { environment, version, namespace } = this;
    const project = this.namespace.project.getValue();

    assert(
      namespace instanceof Namespace,
      'namespace must be an instance of Namespace',
    );
    assert(typeof environment === 'string', 'environment must be a string');
    assert(typeof version === 'string', 'version must be a string');
    assert(typeof project === 'string', 'project value must be a string');
    const uri = `${project}/${environment}/${version}/${this.provider.getUri()}/${namespace.getValue()}`;
    return uri;
  }
}

export default DeploymentConfig;
