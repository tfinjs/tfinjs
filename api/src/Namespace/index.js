import assert from 'assert';
import Project from '../Project/ProjectParent';
import NamespaceParent from './NamespaceParent';

/**
 * Creates an instance of Namespace.
 *
 * @param {project} project - An instance of the Project class
 * @param {string} namespace - The namespace string
 * @class Namespace
 */
class Namespace extends NamespaceParent {
  constructor(project, namespace) {
    super();
    assert(project instanceof Project);
    assert(typeof namespace === 'string', 'namespace must be a string');

    this.namespace = namespace;
    this.project = project;
  }

  /**
   * Gets the value of the namespace
   *
   * @returns {string} namepsace
   * @memberof Namespace
   */
  getValue() {
    return this.namespace;
  }
}

export default Namespace;
