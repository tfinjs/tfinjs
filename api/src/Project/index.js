import fs from 'fs';
import assert from 'assert';
import { isAbsolute } from 'path';
import Backend from '../Backend';
import requiredParam from '../statics/requiredParam';
import Resource from '../Resource';
import resourceExistsInList from '../statics/resourceExistsInList';
import DeploymentConfig from '../DeploymentConfig';
import Namespace from '../Namespace';
import ProjectParent from './ProjectParent';

/**
 * Creates an instance of Project.
 *
 * @param {string} project - The name of the project
 * @param {backend} backen - An instance of the Backend class
 * @param {string} dist - An absolute path pointing to the destination folder for built resources
 * @param {*} [fileSystem=fs]
 * @class Project
 */
class Project extends ProjectParent {
  constructor(project, backend, dist, fileSystem = fs) {
    super();
    assert(typeof project === 'string', 'project must be a string');
    assert(backend instanceof Backend, 'backend must be instance of Backend');

    if (backend.shouldCreateBackend()) {
      /* create dummy deployment and api where the backend can be created */
      const backendBackend = new Backend(null);
      this.backendProject = new Project(
        project,
        backendBackend,
        dist,
        fileSystem,
      );
      const backendNamespace = new Namespace(
        this.backendProject,
        '__tfinjs__backend__',
      );
      const deploymentConfig = new DeploymentConfig(backendNamespace, {
        environment: '_',
        version: '_',
        provider: backend.getProvider(),
      });
      const backendResource = backend.create(deploymentConfig);
      this.addResource(backendResource);
    }

    this.project = project;
    this.backend = backend;

    this.setFs(fileSystem);
    this.setDist(dist);
  }

  /**
   * Gets the filesystem of this project
   *
   * @returns {*} fs
   * @memberof Project
   */
  getFs() {
    return this.fs;
  }

  /**
   * Sets the filesystem of this project
   *
   * @returns {*} fs
   * @memberof Project
   */
  setFs(fileSystem) {
    assert(
      fileSystem.writeFileSync
        && fileSystem.readFileSync
        && fileSystem.mkdir
        && fileSystem.stat,
      'fileSystem must be implemented as the node fs module',
    );
    if (this.backend.shouldCreateBackend()) {
      this.backendProject.setFs(fileSystem);
    }
    this.fs = fileSystem;
  }

  /**
   * Gets the dist folder of this project
   *
   * @returns {string} dist
   * @memberof Project
   */
  getDist() {
    return this.dist;
  }

  /**
   * Updates the dist folder of this project
   *
   * @param {string} dist
   * @memberof Project
   */
  setDist(dist) {
    assert(
      typeof dist === 'string' && isAbsolute(dist),
      'dist must be a string and an absolute path pointing the the dist folder for tfinjs',
    );
    if (this.backend.shouldCreateBackend()) {
      this.backendProject.setDist(dist);
    }
    this.dist = dist;
  }

  /**
   * Gets the name value of this project
   *
   * @returns {string} project - The name of the project
   * @memberof Project
   */
  getValue() {
    return this.project;
  }

  /**
   * Array of added resources
   *
   * @memberof Deployment
   */
  resources = [];

  /**
   * Registers a new resource
   *
   * @param {resource} resource - The resource to be added
   * @memberof Deployment
   */
  addResource(resource = requiredParam('resource')) {
    assert(
      resource instanceof Resource,
      'resource must be an instance of Resource',
    );
    if (resourceExistsInList(this.resources, resource)) {
      const error = new Error(
        'You currently have multiple resources with the same properties which is not allowed',
      );
      throw error;
    }

    this.resources.push(resource);
  }

  /**
   * Gets the resources of the deployment
   *
   * @returns {resources[]} resources - array of resources
   * @memberof Deployment
   */
  getResources() {
    return this.resources;
  }

  /**
   * Gets the dependency graph of the resources of the project
   *
   * @returns {object} graph
   * @memberof Project
   */
  getDependencyGraph() {
    return this.getResources().reduce(
      (nodes, resource) => ({
        ...nodes,
        [resource.versionedName()]: resource
          .getDependencies()
          .map((dependencyResource) => dependencyResource.versionedName()),
      }),
      {},
    );
  }

  /**
   * Gets a resource of the project based on its uri
   *
   * @param {string} uri
   * @returns {resource} resource
   * @memberof Project
   */
  getResourceFromUri(uri) {
    return this.resources.find((resource) => resource.getUri() === uri);
  }

  /**
   * Gets a resource of the project based on its versioned name
   *
   * @param {string} uri
   * @returns {resource} resource
   * @memberof Project
   */
  getResourceFromVersionedName(versionedName) {
    return this.resources.find(
      (resource) => resource.versionedName() === versionedName,
    );
  }

  async build() {
    await Promise.all(
      this.getResources().map(async (resource) => {
        await resource.build();
      }),
    );
  }
}

export default Project;
