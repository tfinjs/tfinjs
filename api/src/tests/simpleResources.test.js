/* eslint-env jest */
import { Volume, createFsFromVolume } from 'memfs';
import { resolve } from 'path';
import awsProviderUri from '../Provider/uris/aws';
import Provider from '../Provider';
import Backend from '../Backend';
import Project from '../Project';
import Namespace from '../Namespace';
import DeploymentConfig from '../DeploymentConfig';
import Resource from '../Resource';
import { reference, versionedName } from '../helpers';

const vol = new Volume();
const fs = createFsFromVolume(vol);

test('simpleResources', async () => {
  const awsAccoundId = 13371337;
  const awsRegion = 'eu-north-1';
  const backendBucketName = 'terraform-state-prod';
  const backendBucketRegion = 'us-east-1';

  const backend = new Backend('s3', {
    backendConfig: (name) => ({
      bucket: backendBucketName,
      key: `${name}.terraform.tfstate`,
      region: backendBucketRegion,
    }),
    dataConfig: (name) => ({
      bucket: backendBucketName,
      key: `${name}.terraform.tfstate`,
      region: backendBucketRegion,
    }),
    provider: new Provider(
      'aws',
      {
        region: backendBucketRegion,
        assume_role: {
          role_arn: `arn:aws:iam::${awsAccoundId}:role/DeploymentRole`,
        },
      },
      awsProviderUri(awsAccoundId, backendBucketRegion),
    ),
    create: (deploymentConfig) =>
      new Resource(deploymentConfig, 'aws_s3_bucket', 'terraform_state_prod', {
        bucket: backendBucketName,
        acl: 'private',
        versioning: {
          enabled: true,
        },
      }),
  });

  const project = new Project('pet-shop', backend, '/', fs);

  const provider = new Provider(
    'aws',
    {
      region: awsRegion,
      assume_role: {
        role_arn: `arn:aws:iam::${awsAccoundId}:role/DeploymentRole`,
      },
    },
    awsProviderUri(awsAccoundId, awsRegion),
  );

  const namespace = new Namespace(project, 'services/lambdas/add-pet');

  const deploymentConfig = new DeploymentConfig(namespace, {
    environment: 'stage',
    version: 'v1',
    provider,
  });

  /* the api is a collection of resources under
   a certain namespace and deployment params. */

  const petLambdaExecRole = new Resource(
    deploymentConfig,
    'aws_iam_role',
    'pets',
    {
      assume_role_policy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Principal: {
              Service: 'lambda.amazonaws.com',
            },
            Effect: 'Allow',
            Sid: '',
          },
        ],
      }),
    },
  );
  const logGroupPrefix = `arn:aws:logs:${awsRegion}:${awsAccoundId}:log-group:/aws/lambda`;
  const petLambda = new Resource(
    deploymentConfig,
    'aws_dynamodb_table',
    'pets',
    {
      description: 'pet lambda',
      role: reference(petLambdaExecRole, 'arn'),
      function_name: versionedName(),
      s3_key: versionedName(),
      s3_bucket: 'pet-lambda-bucket',
      handler: 'service.handler',
      runtime: 'nodejs8.10',
      timeout: 20,
      memory_size: 512,
    },
  );
  const petLambdaName = petLambda.versionedName();
  const cloudwatchPolicy = new Resource(
    deploymentConfig,
    'aws_iam_policy',
    'cloudwatch_attachable_policy',
    {
      policy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: ['logs:CreateLogStream'],
            Effect: 'Allow',
            Resource: `${logGroupPrefix}/${petLambdaName}:*`,
          },
          {
            Action: ['logs:PutLogEvents'],
            Effect: 'Allow',
            Resource: `${logGroupPrefix}/${petLambdaName}:*:*`,
          },
        ],
      }),
    },
  );
  const config = new Resource(
    deploymentConfig,
    'aws_iam_role_policy_attachment',
    'cloud_watch_role_attachment',
    {
      role: reference(petLambdaExecRole, 'name'),
      policy_arn: reference(cloudwatchPolicy, 'arn'),
    },
  );

  await project.build();
  expect(vol.toJSON()).toMatchSnapshot();
});
