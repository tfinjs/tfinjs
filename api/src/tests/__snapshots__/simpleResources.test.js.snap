// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`simpleResources 1`] = `
Object {
  "/tijpetshop191gflg/main.tf": "provider \\"aws\\" {
  assume_role = {
    role_arn = \\"arn:aws:iam::13371337:role/DeploymentRole\\"
  }

  region = \\"eu-north-1\\"
}

terraform {
  backend \\"s3\\" {
    bucket = \\"terraform-state-prod\\"
    key    = \\"tijpetshop191gflg.terraform.tfstate\\"
    region = \\"us-east-1\\"
  }
}

data \\"external\\" \\"save_latest_deploy\\" {
  depends_on = [
    \\"aws_dynamodb_table.pets\\",
  ]

  program = [
    \\"node\\",
    \\"-e\\",
    \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', '3gvkkt')\\",
  ]
}

resource \\"aws_dynamodb_table\\" \\"pets\\" {
  description   = \\"pet lambda\\"
  function_name = \\"tijpetshop191gflg\\"
  handler       = \\"service.handler\\"
  memory_size   = 512

  provisioner \\"local-exec\\" {
    command = \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'DESTROYED')\\"

    interpreter = [
      \\"node\\",
      \\"-e\\",
    ]

    when = \\"destroy\\"
  }

  role      = \\"\${data.terraform_remote_state.tijpetshop1jcp9vz.tfinjs_arn}\\"
  runtime   = \\"nodejs8.10\\"
  s3_bucket = \\"pet-lambda-bucket\\"
  s3_key    = \\"tijpetshop191gflg\\"
  timeout   = 20
}

data \\"terraform_remote_state\\" \\"tijpetshop1jcp9vz\\" {
  backend = \\"s3\\"

  config = {
    bucket = \\"terraform-state-prod\\"
    key    = \\"tijpetshop1jcp9vz.terraform.tfstate\\"
    region = \\"us-east-1\\"
  }
}
",
  "/tijpetshop191gflg/tfinjs.json": "{
  \\"name\\": \\"tijpetshop191gflg\\",
  \\"uri\\": \\"pet-shop/stage/v1/aws/13371337/eu-north-1/services/lambdas/add-pet/aws_dynamodb_table/pets\\",
  \\"contentHash\\": \\"3gvkkt\\",
  \\"dependencies\\": [
    \\"pet-shop/stage/v1/aws/13371337/eu-north-1/services/lambdas/add-pet/aws_iam_role/pets\\"
  ]
}",
  "/tijpetshop1jcp9vz/main.tf": "provider \\"aws\\" {
  assume_role = {
    role_arn = \\"arn:aws:iam::13371337:role/DeploymentRole\\"
  }

  region = \\"eu-north-1\\"
}

terraform {
  backend \\"s3\\" {
    bucket = \\"terraform-state-prod\\"
    key    = \\"tijpetshop1jcp9vz.terraform.tfstate\\"
    region = \\"us-east-1\\"
  }
}

data \\"external\\" \\"save_latest_deploy\\" {
  depends_on = [
    \\"aws_iam_role.pets\\",
  ]

  program = [
    \\"node\\",
    \\"-e\\",
    \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'yxwssf')\\",
  ]
}

resource \\"aws_iam_role\\" \\"pets\\" {
  assume_role_policy = <<EOF
{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Action\\":\\"sts:AssumeRole\\",\\"Principal\\":{\\"Service\\":\\"lambda.amazonaws.com\\"},\\"Effect\\":\\"Allow\\",\\"Sid\\":\\"\\"}]}
EOF

  provisioner \\"local-exec\\" {
    command = \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'DESTROYED')\\"

    interpreter = [
      \\"node\\",
      \\"-e\\",
    ]

    when = \\"destroy\\"
  }
}

output \\"tfinjs_arn\\" {
  value = \\"\${aws_iam_role.pets.arn}\\"
}

output \\"tfinjs_name\\" {
  value = \\"\${aws_iam_role.pets.name}\\"
}
",
  "/tijpetshop1jcp9vz/tfinjs.json": "{
  \\"name\\": \\"tijpetshop1jcp9vz\\",
  \\"uri\\": \\"pet-shop/stage/v1/aws/13371337/eu-north-1/services/lambdas/add-pet/aws_iam_role/pets\\",
  \\"contentHash\\": \\"yxwssf\\",
  \\"dependencies\\": []
}",
  "/tijpetshop92t556/main.tf": "provider \\"aws\\" {
  assume_role = {
    role_arn = \\"arn:aws:iam::13371337:role/DeploymentRole\\"
  }

  region = \\"eu-north-1\\"
}

terraform {
  backend \\"s3\\" {
    bucket = \\"terraform-state-prod\\"
    key    = \\"tijpetshop92t556.terraform.tfstate\\"
    region = \\"us-east-1\\"
  }
}

data \\"external\\" \\"save_latest_deploy\\" {
  depends_on = [
    \\"aws_iam_role_policy_attachment.cloud_watch_role_attachment\\",
  ]

  program = [
    \\"node\\",
    \\"-e\\",
    \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'tiwzqb')\\",
  ]
}

resource \\"aws_iam_role_policy_attachment\\" \\"cloud_watch_role_attachment\\" {
  policy_arn = \\"\${data.terraform_remote_state.tijpetshopkhn8xy.tfinjs_arn}\\"

  provisioner \\"local-exec\\" {
    command = \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'DESTROYED')\\"

    interpreter = [
      \\"node\\",
      \\"-e\\",
    ]

    when = \\"destroy\\"
  }

  role = \\"\${data.terraform_remote_state.tijpetshop1jcp9vz.tfinjs_name}\\"
}

data \\"terraform_remote_state\\" \\"tijpetshop1jcp9vz\\" {
  backend = \\"s3\\"

  config = {
    bucket = \\"terraform-state-prod\\"
    key    = \\"tijpetshop1jcp9vz.terraform.tfstate\\"
    region = \\"us-east-1\\"
  }
}

data \\"terraform_remote_state\\" \\"tijpetshopkhn8xy\\" {
  backend = \\"s3\\"

  config = {
    bucket = \\"terraform-state-prod\\"
    key    = \\"tijpetshopkhn8xy.terraform.tfstate\\"
    region = \\"us-east-1\\"
  }
}
",
  "/tijpetshop92t556/tfinjs.json": "{
  \\"name\\": \\"tijpetshop92t556\\",
  \\"uri\\": \\"pet-shop/stage/v1/aws/13371337/eu-north-1/services/lambdas/add-pet/aws_iam_role_policy_attachment/cloud_watch_role_attachment\\",
  \\"contentHash\\": \\"tiwzqb\\",
  \\"dependencies\\": [
    \\"pet-shop/stage/v1/aws/13371337/eu-north-1/services/lambdas/add-pet/aws_iam_role/pets\\",
    \\"pet-shop/stage/v1/aws/13371337/eu-north-1/services/lambdas/add-pet/aws_iam_policy/cloudwatch_attachable_policy\\"
  ]
}",
  "/tijpetshopkhn8xy/main.tf": "provider \\"aws\\" {
  assume_role = {
    role_arn = \\"arn:aws:iam::13371337:role/DeploymentRole\\"
  }

  region = \\"eu-north-1\\"
}

terraform {
  backend \\"s3\\" {
    bucket = \\"terraform-state-prod\\"
    key    = \\"tijpetshopkhn8xy.terraform.tfstate\\"
    region = \\"us-east-1\\"
  }
}

data \\"external\\" \\"save_latest_deploy\\" {
  depends_on = [
    \\"aws_iam_policy.cloudwatch_attachable_policy\\",
  ]

  program = [
    \\"node\\",
    \\"-e\\",
    \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'kz9og5')\\",
  ]
}

resource \\"aws_iam_policy\\" \\"cloudwatch_attachable_policy\\" {
  policy = <<EOF
{\\"Version\\":\\"2012-10-17\\",\\"Statement\\":[{\\"Action\\":[\\"logs:CreateLogStream\\"],\\"Effect\\":\\"Allow\\",\\"Resource\\":\\"arn:aws:logs:eu-north-1:13371337:log-group:/aws/lambda/tijpetshop191gflg:*\\"},{\\"Action\\":[\\"logs:PutLogEvents\\"],\\"Effect\\":\\"Allow\\",\\"Resource\\":\\"arn:aws:logs:eu-north-1:13371337:log-group:/aws/lambda/tijpetshop191gflg:*:*\\"}]}
EOF

  provisioner \\"local-exec\\" {
    command = \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'DESTROYED')\\"

    interpreter = [
      \\"node\\",
      \\"-e\\",
    ]

    when = \\"destroy\\"
  }
}

output \\"tfinjs_arn\\" {
  value = \\"\${aws_iam_policy.cloudwatch_attachable_policy.arn}\\"
}
",
  "/tijpetshopkhn8xy/tfinjs.json": "{
  \\"name\\": \\"tijpetshopkhn8xy\\",
  \\"uri\\": \\"pet-shop/stage/v1/aws/13371337/eu-north-1/services/lambdas/add-pet/aws_iam_policy/cloudwatch_attachable_policy\\",
  \\"contentHash\\": \\"kz9og5\\",
  \\"dependencies\\": []
}",
  "/tijpetshopvj8fsk/main.tf": "provider \\"aws\\" {
  assume_role = {
    role_arn = \\"arn:aws:iam::13371337:role/DeploymentRole\\"
  }

  region = \\"us-east-1\\"
}

data \\"external\\" \\"save_latest_deploy\\" {
  depends_on = [
    \\"aws_s3_bucket.terraform_state_prod\\",
  ]

  program = [
    \\"node\\",
    \\"-e\\",
    \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', '1mxul7o')\\",
  ]
}

resource \\"aws_s3_bucket\\" \\"terraform_state_prod\\" {
  acl    = \\"private\\"
  bucket = \\"terraform-state-prod\\"

  provisioner \\"local-exec\\" {
    command = \\"require('@tfinjs/api').utils.saveDeploymentStatus('\${path.root}', 'DESTROYED')\\"

    interpreter = [
      \\"node\\",
      \\"-e\\",
    ]

    when = \\"destroy\\"
  }

  versioning = {
    enabled = true
  }
}
",
  "/tijpetshopvj8fsk/tfinjs.json": "{
  \\"name\\": \\"tijpetshopvj8fsk\\",
  \\"uri\\": \\"pet-shop/_/_/aws/13371337/us-east-1/__tfinjs__backend__/aws_s3_bucket/terraform_state_prod\\",
  \\"contentHash\\": \\"1mxul7o\\",
  \\"dependencies\\": []
}",
}
`;
